File: src/Relationship.java
import edu.stanford.nlp.coref.data.CorefChain;
import edu.stanford.nlp.ling.CoreLabel;
import edu.stanford.nlp.pipeline.CoreDocument;
import edu.stanford.nlp.pipeline.CoreSentence;
import edu.stanford.nlp.pipeline.StanfordCoreNLP;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Created by liuziqi on 2018/4/15.
 */
public class Relationship {
    private final Set<String> pronoun = Stream.of("he", "she", "it", "him", "her", "they", "them").collect(Collectors.toSet());

    private final String PERSON = "PERSON"; // NER tag, can also add ORGANIZATION if needed
    private int entityCount; // total entity, also as an increasing ID, but notice that some entities might be merged, causing the missing id
    private int mergeCount;
    private int threshold;  // how many words allowed between two entities that they can be recognized as relationship
    StanfordCoreNLP pipeline;
    private boolean debug;
    ArrayList<String> sections;  // split the whole text into sections according to different granularity
    HashMap<String, Entity> entityNameMap; // the same entity might have different names
    HashSet<Entity> entitySet;


    public Relationship(StanfordCoreNLP pipeline, String text, String granularity, int threshold, boolean debug) {
        this.debug = debug;
        this.entityCount = 0;
        this.mergeCount = 0;
        this.threshold = threshold;
        this.pipeline = pipeline;
        this.entityNameMap = new HashMap<>();
        this.entitySet = new HashSet<>();
        sections = getSections(text, granularity);
        if (debug) {
            System.out.println("document divided into " + this.sections.size() + " sections");
        }

        for (int i = 0; i < sections.size(); i++) {
            // for each section
            String section = sections.get(i);
            if (debug)
                System.out.println("processing section " + i);
            List<MentionToken> tokens = this.parseCoref(section);
            parseRelationship(tokens);
        }
    }

    private ArrayList<String> getSections(String text, String granularity) {
        return Arrays.stream(text.trim().split("\\n{2,}")).collect(Collectors.toCollection(ArrayList::new));
    }

    /**
     * @param tokens mention tokens from one section
     */
    private void parseRelationship(List<MentionToken> tokens) {
        System.out.println(Arrays.toString(entityNameMap.keySet().toArray()));
        // Noticed that the tokens appear in their text order
        // Noticed that all overlapping entities have already been merged here,
        // so overlapping entities won't interact with each other
        Queue<MentionToken> queue = new LinkedList<>();
        for (MentionToken m : tokens) {

            // remove those far away entity. (not in the same sentence and [threshold words] aways
            while (!queue.isEmpty() && m.absPos - queue.peek().absPos > threshold && queue.peek().senNum < m.senNum) {
                queue.poll();
            }


            for (Entity neighbor : queue.stream().map(o -> o.entity).distinct().collect(Collectors.toList())) {
                happenRelationship(m.entity, neighbor);
            }

            queue.offer(m);
        }
    }

    private void happenRelationship(Entity e1, Entity e2) {
        // 发生关系！
        if (!e1.equals(e2)) {
            e1.neighborInteract.put(e2, e1.neighborInteract.getOrDefault(e2, 0) + 1);
            e2.neighborInteract.put(e1, e2.neighborInteract.getOrDefault(e1, 0) + 1);
        }

    }


    /**
     * @param text original text from one section
     * @return list of mention token (sorted in their textual order)
     */
    private List<MentionToken> parseCoref(String text) {

        // a big problem is that coref annotator will skip those named entity which is not referred by a pronoun
        // for example "David is a boy" won't produce any result from coref annotator, while "David is a boy and he like swimming" will.
        // therefore, we should also use NER combined with Coref result

        // this is useful when parsing NER, to avoid duplicate entry from Coref, identified as absPos
        HashMap<Integer, MentionToken> tokenMap = new HashMap<>(); // key: absPos

        CoreDocument document = new CoreDocument(text);
        pipeline.annotate(document);

        if (debug) {
            System.out.println(document.corefChains().values().size() + " coref chains detected.");
            for (CorefChain cc : document.corefChains().values()) {
                System.out.println(cc);
            }
        }
        List<MentionToken> mentions = new ArrayList<>();

        // accumulated sum. the start index (in the whole text) of each sentence
        List<Integer> sentenceLen = Stream.of(0).collect(Collectors.toList());
        document.sentences().forEach(o -> sentenceLen.add(sentenceLen.get(sentenceLen.size() - 1) + o.tokens().size()));


        for (CorefChain corefChain : document.corefChains().values()) {
            List<CorefChain.CorefMention> cluster = corefChain.getMentionsInTextualOrder();
            Entity entity = getEntity(cluster);

            for (CorefChain.CorefMention m : cluster) {
                MentionToken token = new MentionToken(
                        m.mentionID,
                        m.startIndex + sentenceLen.get(m.sentNum - 1) - 1,
                        m.sentNum, m.corefClusterID,
                        m.mentionSpan.toLowerCase(),
                        entity);
                mentions.add(token);
                tokenMap.put(token.absPos, token);
            }
        }

        // parse NER here
        int maxClusterID = mentions.stream().map(o -> o.clusterID).max((o1, o2) -> o1 - o2).orElse(0);
        // we need to manually assign clusterID to those entity that doesn't appear in Coref
        List<CoreSentence> sentences = document.sentences();
        for (int i = 0; i < sentences.size(); i++) {
            // iterate over sentences
            CoreSentence sen = sentences.get(i);
            List<CoreLabel> tokens = sen.tokens();
            int startIdx = -1;
            boolean pre = false;
            for (int j = 0; j < tokens.size(); j++) {
                // iterate over tokens
                CoreLabel tk = tokens.get(j);
                if (tk.ner().equals(PERSON)) {
                    if (!pre) {
                        pre = true;
                        startIdx = j;
                    }
                    if (j == tokens.size() - 1) {
                        int endIdx = j;
                        int absPos = sentenceLen.get(i) + startIdx;
                        if (!tokenMap.containsKey(absPos))
                            extractNerEntity(startIdx, endIdx, text, tokens, mentions, i, absPos, ++maxClusterID);
                    }
                } else {
                    if (pre) {
                        pre = false;
                        int endIdx = j - 1;
                        int absPos = sentenceLen.get(i) + startIdx;
                        if (!tokenMap.containsKey(absPos))
                            extractNerEntity(startIdx, endIdx, text, tokens, mentions, i, absPos, ++maxClusterID);
                    }

                }
            }

        }

        // sorted by their order in the text
        mentions.sort((Comparator.comparingInt(o -> o.absPos)));
        if (debug) {
            for (MentionToken m : mentions) {
                System.out.println(m.name + " " + m.absPos);
            }
        }
        return mentions;
    }

    /**
     * a helper function called in parseCoref
     *
     * @param startIdx
     * @param endIdx
     * @param text
     * @param tokens
     * @param mentions
     * @param senNum
     * @param absPos
     * @param id
     */
    private void extractNerEntity(int startIdx, int endIdx, String text, List<CoreLabel> tokens, List<MentionToken> mentions, int senNum, int absPos, int id) {

        String name = text.substring(tokens.get(startIdx).beginPosition(), tokens.get(endIdx).endPosition()).toLowerCase();
        Entity entity = entityNameMap.getOrDefault(name, new Entity(entityCount++));
        entity.frequency++;
        entity.names.add(name);

        mentions.add(new MentionToken(-1, absPos, senNum, id, name, entity));
        entityNameMap.put(name, entity);
        entitySet.add(entity);
    }

    /**
     * @param cluster a coref cluster
     * @return identify the entity they refer to (if non-exist then create a new one)
     */
    private Entity getEntity(List<CorefChain.CorefMention> cluster) {
        int appearCount = cluster.size();
        List<String> cluster_names = cluster.stream()
                .map(o -> o.mentionSpan.toLowerCase())
                .distinct()
                .filter(o -> !pronoun.contains(o))
                .collect(Collectors.toList());

        // if all names are pronoun (but I suppose this situation will not happen
        if (cluster_names.isEmpty())
            return null;

        List<Entity> entities = cluster_names.stream()
                .map(o -> entityNameMap.getOrDefault(o, null))
                .distinct()
                .filter(Objects::nonNull)
                .collect(Collectors.toList());

        // if zero entity, which mean this entity hans't appear before
        // one entity, that's the best case
        // more than one entity, we merge those entities here
        Entity entity = null;
        switch (entities.size()) {
            case 0:
                entity = new Entity(entityCount++);
                break;
            case 1:
                entity = entities.get(0);
                break;
            default:
                // we need to merge entities here
                this.mergeCount += entities.size() - 1;
                Entity merge = entities.get(0);
                for (int i = 1; i < entities.size(); i++) {
                    Entity temp = entities.get(i);
                    // remove from entitySet
                    this.entitySet.remove(temp);
                    // merge names
                    merge.names.addAll(temp.names);
                    // merge frequency
                    merge.frequency += temp.frequency;
                    // merge neighbor
                    for (Map.Entry<Entity, Integer> nb : temp.neighborInteract.entrySet()) {
                        if (merge.neighborInteract.containsKey(nb.getKey())) {
                            merge.neighborInteract.put(nb.getKey(), merge.neighborInteract.get(nb.getKey()) + nb.getValue());
                        }
                    }
                }
                // delete itself from neighbors
                entities.forEach(e -> merge.neighborInteract.remove(e));
                entity = merge;
                break;
        }

        entitySet.add(entity);
        entity.names.addAll(cluster_names);
        for (String name : cluster_names) {
            entityNameMap.put(name, entity);
        }

        entity.frequency += appearCount;
        return entity;
    }


    public int entityNumber() {
        return this.entityCount - mergeCount;
    }

    public String report() {
        StringBuilder sb = new StringBuilder("Total entity numbers: " + this.entityNumber() + "\n\n");
        List<Entity> entityList = entitySet.stream().sorted((o1, o2) -> o2.frequency - o1.frequency).collect(Collectors.toList());

        sb.append("Entities sorted by frequency:").append("\n");
        entityList.forEach(o -> sb.append(String.format("%s\t%d\n", String.join(",", o.names), o.frequency)));

        sb.append("\n\nInteraction\n");
        for (Entity e : entityList) {
            sb.append("===========================\n");
            sb.append(e.toString()).append("\n").append("-----------------\n");
            for (Map.Entry<Entity, Integer> entry : e.neighborInteract.entrySet()) {
                sb.append(entry.getKey().toString()).append("\t").append(entry.getValue()).append("\n");
            }
        }

        return sb.toString();
    }
}

class Entity {
    public int id;
    public Set<String> names;
    public int frequency;
    public HashMap<Entity, Integer> neighborInteract;

    public Entity(int id) {
        this.id = id;
        this.names = new HashSet<>();
        this.frequency = 0;
        neighborInteract = new HashMap<>();
    }

    public String toString() {
        return Arrays.toString(names.toArray());
    }
}


class MentionToken {
    public int mentionID;
    public int absPos;
    public int senNum;
    public int clusterID;
    public String name;
    public Entity entity;

    public MentionToken(int mentionID, int absPos, int senNum, int clusterID, String name, Entity entity) {
        this.mentionID = mentionID;
        this.absPos = absPos;
        this.senNum = senNum;
        this.clusterID = clusterID;
        this.name = name;
        this.entity = entity;
    }

}

File: pysrc/relationship.py
# encoding: utf-8

'''

@author: ZiqiLiu


@file: relationship.py

@time: 2018/4/16 下午11:23

@desc:
'''
import numpy as np
from module import Entity, Token
from collections import deque
import re


class Relationship(object):
    pronoun = ["he", "she", "it", "him", "her", "they", "them", "this", "that"]
    PERSON = 1

    def __init__(self, pipeline, text, threshold=25, debug=True):
        pat = re.compile(r'\n+')
        self.ner = set()
        self.mergeCount = 0
        self.debug = debug
        self.entityCount = 0
        self.threshold = threshold
        self.pipeline = pipeline
        self.entityMap = {}
        # self.entitySet = set()
        text = pat.sub(' ', text)

        tokens = self.parseCoref(text)
        self.parseRelationship(tokens)

    def parseRelationship(self, tokens):
        # Noticed that the tokens are sorted in their text order
        queue = deque()
        for m in tokens:
            while len(queue) > 0 and m.absPos - queue[
                0].absPos > self.threshold:
                queue.popleft()
            for nb in set(queue):
                self._happenRelationship(m.entity, nb.entity)
            queue.append(m)

    def _happenRelationship(self, e1, e2):
        '''
        
        :param e1: Entity 
        :param e2: Entity
        :return: 
        '''
        if e1 != e2:
            e1.neighbors[e2] = e1.neighbors.get(e2, 0) + 1
            e2.neighbors[e1] = e2.neighbors.get(e1, 0) + 1

    def parseCoref(self, text):
        tokens = []

        self.pipeline.continuous_coref(utterances=text)
        doc = self.pipeline.get_utterances()[0]
        doc_text = [d.text.strip() for d in doc]

        filtered_ner = list(
            filter(lambda x: self._NERfilter(x), doc.ents))
        ner = np.zeros((len(doc)), dtype=np.int32)
        self.ner.update([str(e) for e in filtered_ner])
        for ent in filtered_ner:
            ner[ent.start:ent.end] = 1

        # clusters id is corresponded to mentions
        clusters = self.pipeline.get_clusters(remove_singletons=True)[
            0]  # ({1: [1], [1]}
        mentions = self.pipeline.get_mentions()
        # this is useful when parsing NER, to avoid duplicate entry from Coref, identified as absPos
        token_map = {}

        if self.debug:
            print('%d chain detected.' % len(clusters))
            name_clusters = [
                [(mentions[idx].text, mentions[idx].start) for idx in cluster]
                for cluster
                in clusters.values()]

            for cluster in name_clusters:
                print(cluster)

        def name_filter(m):
            '''
            check how many entity in this mention token, and cut the name to entity name
            :param m: 
            :return: a tuple of (m, modify_name, absPos,display_name)
            '''

            ents_span = []
            pre = False
            start_idx = -1
            name = None
            for j in range(m.start, m.end):
                if ner[j] == self.PERSON:
                    if not pre:
                        start_idx = j
                        pre = True
                    if j == m.end - 1:
                        ents_span.append((start_idx, j + 1))
                else:
                    if pre:
                        pre = False
                        ents_span.append((start_idx, j))
            if len(ents_span) > 1:
                # more than one entity
                print('fuck!')
                print(m.text)
            elif len(ents_span) == 1:
                st, end = ents_span[0]
                name = ' '.join(doc_text[st:end]).lower()
            display_name = name if name is not None else m.text
            return m, name, ents_span[0][0] if len(
                ents_span) > 0 else m.start, display_name

        # parse Coref
        if self.debug:
            print('\n\nparsing coref.....\n')
        for _, chain in clusters.items():
            # chain is only id for mention tokens
            ms = [mentions[key] for key in chain]

            ms_tuple = [name_filter(m) for m in ms]
            names = list(
                filter(lambda x: x is not None, map(lambda x: x[1], ms_tuple)))
            if len(names) == 0:
                if self.debug:
                    print('empty chain, skip', ms)
                continue
            if self.debug:
                print(names, ms)
            ent = self._getEntity_coref(names=names, count=len(ms))

            for m, modified_name, absPos, display_name in ms_tuple:
                token = Token(absPos, display_name, ent)
                tokens.append(token)
                token_map[absPos] = token

        # parse NER

        start_idx = -1
        pre = False
        for i in range(len(ner)):
            if ner[i] == self.PERSON:
                if not pre:
                    pre = True
                    start_idx = i
                if i == len(ner) - 1:
                    # end of array
                    if start_idx not in token_map:
                        self._getEntity_NER(start_idx=start_idx, end_idx=i + 1,
                                            doc_text=doc_text, tokens=tokens)
            else:
                if pre:
                    pre = False
                    if start_idx not in token_map:
                        self._getEntity_NER(start_idx=start_idx, end_idx=i,
                                            doc_text=doc_text, tokens=tokens)

        tokens.sort(key=lambda x: x.absPos)
        if self.debug:
            print("\n\ntokens sorted in text order:")
            for tk in tokens:
                print("%s\t%d" % (tk.name, tk.absPos))
        return tokens

    def _getEntity_coref(self, names, count):

        ent = None

        existing_ents = [self.entityMap[name] for name in names if
                         name in self.entityMap]

        if len(existing_ents) == 1:
            ent = existing_ents[0]
        elif len(existing_ents) > 1:
            # hopefully we won't meet this case
            print('merge!!!', [e.names for e in existing_ents])
            # merge entities
            self.mergeCount += len(existing_ents) - 1
            merge = existing_ents[0]
            for to_merge in existing_ents[1:]:
                # remove from entitySet
                # self.entitySet.remove(to_merge)
                # merge freq
                merge.freq += to_merge.freq
                # merge names
                merge.names.update(to_merge.names)
                # merge neighbor
                for nb, ct in to_merge.neighbors.items():
                    if nb != merge:
                        merge.neighbors[nb] = merge.neighbors.get(nb, 0) + ct
                        nb.neighbors[merge] = nb.neighbors.get(merge, 0) + ct
                        nb.neighbors.pop(to_merge)
                    else:
                        merge.neighbors.pop(to_merge)
            ent = merge
        else:
            # create new one
            self.entityCount += 1
            ent = Entity(self.entityCount)

        # self.entitySet.add(ent)
        ent.names.update(names)
        ent.freq += count
        for name in ent.names:
            self.entityMap[name] = ent
        return ent

    def _getEntity_NER(self, start_idx, end_idx, doc_text, tokens):
        name = ' '.join(doc_text[start_idx:end_idx]).lower()
        if name in self.entityMap:
            ent = self.entityMap[name]
        else:
            self.entityCount += 1
            ent = Entity(self.entityCount)

        ent.freq += 1
        ent.names.add(name)
        # self.entitySet.add(ent)
        self.entityMap[name] = ent
        tokens.append(Token(absPos=start_idx, name=name, entity=ent))

    def _NERfilter(self, e):
        words = e.text.split()
        return any(w[0].isupper() for w in e.text.split())

    @property
    def totalEntityNum(self):
        return self.entityCount - self.mergeCount

    def __str__(self):
        report = ''
        sortedEntity = sorted(set(self.entityMap.values()),
                              key=lambda x: x.freq,
                              reverse=True)
        report += "Entities sorted by frequency:\n"
        for ent in sortedEntity:
            report += "%s\t%d\n" % (ent, ent.freq)

        report += '\n\nRelationship\n'
        for ent in sortedEntity:
            report += '=' * 30 + '\n'
            report += str(ent) + '\n'
            report += '-' * 20 + '\n'
            for nb, ct in ent.neighbors.items():
                report += '%s\t%d\n' % (nb, ct)
        return report

    def report(self):
        with open('./ner.txt', 'w') as f:
            for ner in self.ner:
                f.write(ner + '\n')
        print(self)
        print(self.ner)
